### PR TITLE
Sort the server cert chain before verify.

### PR DESCRIPTION
**Pull request by ArthurMa1978**
Notes:
* Pull requests cannot be accepted until:
-  The submitter has [accepted the online agreement here with a click through](https://developer.mbed.org/contributor_agreement/) 
   or for companies or those that do not wish to create an mbed account, a slightly different agreement can be found [here](https://www.mbed.com/en/about-mbed/contributor-license-agreements/)
- The PR follows the [mbed TLS coding standards](https://tls.mbed.org/kb/development/mbedtls-coding-standards)
* This is just a template, so feel free to use/remove the unnecessary things
## Description
Sort the server cert chain to handle potentially extraneous certificates and arbitrary orderings.
This is the fix for https://github.com/ARMmbed/mbed-os/issues/9868.

## Status
**READY/IN DEVELOPMENT/HOLD**

## Requires Backporting
When there is a bug fix, it should be backported to all maintained and supported branches.
Changes do not have to be backported if:
- This PR is a new feature\enhancement
- This PR contains changes in the API. If this is true, and there is a need for the fix to be backported, the fix should be handled differently in the legacy branch

NO  

## Migrations
If there is any API change, what's the incentive and logic for it.

NO

## Additional comments
Already had a pr https://github.com/ARMmbed/mbed-os/pull/9909 in mbed-os repo.
Since there have huge different between mbed-os/feature/mbedtls and this repo, I don't know which one should be the right place that I should send the pr.

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Changelog updated
- [ ] Backported

## Steps to test or reproduce
N/A

### CODE REVIEW

**Comment by hanno-arm** [comment]
Hi @ArthurMa1978,

thank you for your report and contribution! Also, thank you for pointing out that TLS 1.3 relaxes the constraints where TLS 1.2 did required strict ordering. I was aware of the latter, but not of the former. 

The current chain verification in Mbed TLS goes midway: We don't require that the `n`-th CRT in the chain is signed by the `n+1`-th CRT, as mandated by the standard, but only that it's signed by the `m`-th CRT with `m > n`. The core of this search for parents is `x509_crt_find_parent_in()`, see https://github.com/ARMmbed/mbedtls/blob/development/library/x509_crt.c#L2052, where `child` takes the role of _both_ the CRT to find a parent for, as well as the remaining chain to search in. 

Currently, I am wondering whether a cheaper (in code-size) approach to fixing your problem is to modify `x509_crt_find_parent_in()` to always search the _entire_ CRT chain provided by the peer. This would require only minimal changes to the code. However, it needs to be considered whether this might lead to undesired looping behavior in case of self-signed certificates within the chain.

I haven't looked at your code in detail yet, but on first sight, there might still be a problem with self-issued (not self-signed) roll-over CRTs in the peer's chain: Those have the same subject name, so a subject-based sorting function doesn't necessarily change their order, but the current verification algorithm _only_ works if the self-issued rollover CRT precedes the updated CRT.

These are just some first thoughts - I'll try to find time to look closer and come back to you then. 

cc @mpg 

Kind regards, and thank you again for your contribution,
Hanno

**Comment by RonEld** [comment]
Hi @ArthurMa1978 Thank you fore your contribution!

Is this contribution a private contribution, or as part of your work at Microsoft?

**Comment by ArthurMa1978** [comment]
> Hi @ArthurMa1978,
> thank you for your report and contribution! Also, thank you for pointing out that TLS 1.3 relaxes the constraints where TLS 1.2 did required strict ordering. I was aware of the latter, but not of the former.
> The current chain verification in Mbed TLS goes midway: We don't require that the n-th CRT in the chain is signed by the n+1-th CRT, as mandated by the standard, but only that it's signed by the m-th CRT with m > n. The core of this search for parents is x509_crt_find_parent_in(), see https://github.com/ARMmbed/mbedtls/blob/development/library/x509_crt.c#L2052, where child takes the role of both the CRT to find a parent for, as well as the remaining chain to search in.
> Currently, I am wondering whether a cheaper (in code-size) approach to fixing your problem is to modify x509_crt_find_parent_in() to always search the entire CRT chain provided by the peer. This would require only minimal changes to the code. However, it needs to be considered whether this might lead to undesired looping behavior in case of self-signed certificates within the chain.
> I haven't looked at your code in detail yet, but on first sight, there might still be a problem with self-issued (not self-signed) roll-over CRTs in the peer's chain: Those have the same subject name, so a subject-based sorting function doesn't necessarily change their order, but the current verification algorithm only works if the self-issued rollover CRT precedes the updated CRT.
> These are just some first thoughts - I'll try to find time to look closer and come back to you then.
> cc @mpg
> Kind regards, and thank you again for your contribution,
> Hanno

Thanks @hanno-arm.
I agree that ```x509_crt_find_parent_in``` might be the better place for this fixing, actually I that's my initial idea but as I'm not a mbedtls expert so choose the much safer way.
And for self-issued cert, hmm... that's what I did aware, but here the sort function will not change anything because it just stop sorting if the cert is self-issued.

**Comment by ArthurMa1978** [comment]
> Hi @ArthurMa1978 Thank you fore your contribution!
> Is this contribution a private contribution, or as part of your work at Microsoft?

Hi @RonEld , I met this issue in one project I'm working on. I can workaround it, but should contribute back in a open source way :)

**Comment by RonEld** [comment]
@ArthurMa1978 Thank you for your confirmation.
From your answer, I understand this is a corporate contribution, so using the accepted CLA from your organiztion.

**Comment by ArthurMa1978** [comment]
any update? I saw the CI was failed

**Comment by RonEld** [comment]
@ArthurMa1978 
The CI failed because of the following checks:
~~~
$ tests/scripts/check-files.py
Trailing whitespace:
./library/x509_crt.c: 2658, 2659, 2660, 2661, 2662, 2664
The command "tests/scripts/check-files.py" exited with 1.
7.13s$ tests/scripts/doxygen.sh
/home/travis/build/ARMmbed/mbedtls/include/mbedtls/x509_crt.h:505: warning: argument 'crt' of command @param is not found in the argument list of mbedtls_x509_crt_sort(mbedtls_x509_crt *chain)
/home/travis/build/ARMmbed/mbedtls/include/mbedtls/x509_crt.h:505: warning: The following parameters of mbedtls_x509_crt_sort(mbedtls_x509_crt *chain) are not documented:
FAIL
The command "tests/scripts/doxygen.sh" exited with 1.
~~~

**Comment by ArthurMa1978** [comment]
> @ArthurMa1978
> The CI failed because of the following checks:
> $ tests/scripts/check-files.py
> Trailing whitespace:
> ./library/x509_crt.c: 2658, 2659, 2660, 2661, 2662, 2664
> The command "tests/scripts/check-files.py" exited with 1.
> 7.13s$ tests/scripts/doxygen.sh
> /home/travis/build/ARMmbed/mbedtls/include/mbedtls/x509_crt.h:505: warning: argument 'crt' of command @param is not found in the argument list of mbedtls_x509_crt_sort(mbedtls_x509_crt *chain)
> /home/travis/build/ARMmbed/mbedtls/include/mbedtls/x509_crt.h:505: warning: The following parameters of mbedtls_x509_crt_sort(mbedtls_x509_crt *chain) are not documented:
> FAIL
> The command "tests/scripts/doxygen.sh" exited with 1.

@RonEld Fixed

**Comment by ArthurMa1978** [comment]
@RonEld  I saw the Jenkins CI failed, it's your internal system so no idea what's happened, can you help on it?

File: include/mbedtls/x509_crt.h [hunk 1]
```diff
@@ -501,6 +501,22 @@ int mbedtls_x509_crt_verify_restartable( mbedtls_x509_crt *crt,
                      void *p_vrfy,
                      mbedtls_x509_crt_restart_ctx *rs_ctx );
 
+/**
+ * \brief          Sort the server cert chain to handle potentially extraneous certificates and arbitrary orderings.
```
Comment (hanno-arm) [inline include/mbedtls/x509_crt.h]: Style: Please avoid lines longer than 80 characters.

File: include/mbedtls/x509_crt.h [hunk 2]
```diff
@@ -501,6 +501,22 @@ int mbedtls_x509_crt_verify_restartable( mbedtls_x509_crt *crt,
...
+ *                 Prior to TLS 1.3, "certificate_list" ordering required each certificate to certify the one immediately
+ *                 preceding it; however, some implementations allowed some flexibility. Servers sometimes send both a
+ *                 current and deprecated intermediate for transitional purposes, and others are simply configured incorrectly,
+ *                 but these cases can nonetheless be validated properly. For maximum compatibility,
+ *                 all implementations SHOULD be prepared to handle potentially extraneous certificates and arbitrary orderings
+ *                 from any TLS version, with the exception of the end-entity certificate which MUST be first.
+ *
+ * \param chain    a certificate chain to be sorted
+ */
+void mbedtls_x509_crt_sort( mbedtls_x509_crt *chain );
```
Comment (hanno-arm) [inline include/mbedtls/x509_crt.h]: What's the rationale for making this part of the public API? I'd expect it to be an internal function only.

File: library/ssl_tls.c [hunk 3]
```diff
@@ -6057,6 +6057,7 @@ static int ssl_parse_certificate_verify( mbedtls_ssl_context *ssl,
     /*
      * Main check: verify certificate
      */
+    mbedtls_x509_crt_sort(chain);
```
Comment (hanno-arm) [inline library/ssl_tls.c]: Minor: Style.

File: library/x509_crt.c [hunk 4]
```diff
@@ -2653,6 +2653,60 @@ int mbedtls_x509_crt_verify_restartable( mbedtls_x509_crt *crt,
...
+ * all implementations SHOULD be prepared to handle potentially extraneous certificates and arbitrary orderings
+ * from any TLS version, with the exception of the end-entity certificate which MUST be first.
+ *
+ * Sort the server cert chain o handle potentially extraneous certificates and arbitrary orderings.
+ */
+void mbedtls_x509_crt_sort( mbedtls_x509_crt *chain )
+{
+    mbedtls_x509_crt *child = chain;
+
+    while (child)
```
Comment (hanno-arm) [inline library/x509_crt.c]: Minor: Please switch to Mbed TLS' coding style.

File: include/mbedtls/x509_crt.h (hunk 1 above)
Comment (ArthurMa1978) [reply include/mbedtls/x509_crt.h]: Fixed.

File: include/mbedtls/x509_crt.h (hunk 2 above)
Comment (ArthurMa1978) [reply include/mbedtls/x509_crt.h]: This function is more like a utility for x509 certificate parsing, so I think put it in x509_crt.h/c and make it as a public API is straight forward, even only used by ssl_tls.

**Comment by ArthurMa1978** [comment]
Thanks @hanno-arm , I have fixed all style comments. For the `public` question please see my comments above.

File: include/mbedtls/x509_crt.h (hunk 2 above)
Comment (hanno-arm) [reply include/mbedtls/x509_crt.h]: @ArthurMa1978 We can always make a function public if we feel it's of general use, but we cannot easily undo it, so it needs to be carefully considered. In this case, I don't see a strong reason to introduce this function to the public API. Could you please make it internal?

File: include/mbedtls/x509_crt.h [hunk 5]
```diff
@@ -526,6 +526,7 @@ int mbedtls_x509_crt_verify_restartable( mbedtls_x509_crt *crt,
                      mbedtls_x509_crt_restart_ctx *rs_ctx );
 
 /**
+<<<<<<< HEAD
```
Comment (hanno-arm) [inline include/mbedtls/x509_crt.h]: Rebase artefact.

File: include/mbedtls/x509_crt.h [hunk 6]
```diff
@@ -591,6 +592,18 @@ int mbedtls_x509_crt_verify_with_ca_cb( mbedtls_x509_crt *crt,
                      void *p_vrfy );
 
 #endif /* MBEDTLS_X509_TRUSTED_CERTIFICATE_CALLBACK */
+=======
```
Comment (hanno-arm) [inline include/mbedtls/x509_crt.h]: Rebase artefact.

File: .gitignore [hunk 7]
```diff
@@ -26,3 +26,6 @@ massif-*
 
 # CMake generates *.dir/ folders for in-tree builds (used by MSVC projects), ignore all of those:
 *.dir/
+
```
Comment (hanno-arm) [inline .gitignore]: I'm sorry, could you please give the rationale for this change? It doesn't seem to be related to the issue this PR sets out to solve.

File: include/mbedtls/x509_crt.h [hunk 8]
```diff
@@ -560,6 +559,20 @@ typedef int (*mbedtls_x509_crt_ca_cb_t)( void *p_ctx,
                                          mbedtls_x509_crt const *child,
                                          mbedtls_x509_crt **candidate_cas );
 
+/*
+ * \brief       Sort the server cert chain to handle potentially extraneous
+ *              certificates and arbitrary orderings.
+ *              For more detail pelease check:
```
Comment (hanno-arm) [inline include/mbedtls/x509_crt.h]: Typo: `pelease` -> `please`

File: include/mbedtls/x509_crt.h [hunk 9]
```diff
@@ -560,6 +559,20 @@ typedef int (*mbedtls_x509_crt_ca_cb_t)( void *p_ctx,
...
+ *           https://tools.ietf.org/id/draft-ietf-tls-tls13-23.html#certificate.
+ *
+ * \note        Performs the same job as \c mbedtls_crt_verify_with_profile()
```
Comment (hanno-arm) [inline include/mbedtls/x509_crt.h]: This seems to be a copy-paste mistake from the documentation of `mbedtls_crt_verify_xxx()`. Could you please remove it?

File: include/mbedtls/x509_crt.h (hunk 5 above)
Comment (ArthurMa1978) [reply include/mbedtls/x509_crt.h]: fixed

File: include/mbedtls/x509_crt.h (hunk 6 above)
Comment (ArthurMa1978) [reply include/mbedtls/x509_crt.h]: fixed

File: .gitignore (hunk 7 above)
Comment (ArthurMa1978) [reply .gitignore]: Yes, not related to this issue, but just for vscode user.
Will revert it.

File: library/x509_crt.c (hunk 4 above)
Comment (ArthurMa1978) [reply library/x509_crt.c]: fixed

File: library/ssl_tls.c (hunk 3 above)
Comment (ArthurMa1978) [reply library/ssl_tls.c]: fixed

File: include/mbedtls/x509_crt.h (hunk 2 above)
Comment (ArthurMa1978) [reply include/mbedtls/x509_crt.h]: Make sense, fixed.

File: .gitignore (hunk 7 above)
Comment (ArthurMa1978) [reply .gitignore]: fixed

File: include/mbedtls/x509_crt.h (hunk 9 above)
Comment (ArthurMa1978) [reply include/mbedtls/x509_crt.h]: fixed.

File: include/mbedtls/x509_crt.h (hunk 8 above)
Comment (ArthurMa1978) [reply include/mbedtls/x509_crt.h]: fixed

File: library/x509_crt.c [hunk 10]
```diff
@@ -2681,6 +2681,62 @@ static int x509_crt_verify_restartable_ca_cb( mbedtls_x509_crt *crt,
...
+ * properly. For maximum compatibility, all implementations SHOULD be prepared
+ * to handle potentially extraneous certificates and arbitrary orderings from
+ * any TLS version, with the exception of the end-entity certificate which
+ * MUST be first.
+ */
+static void mbedtls_x509_crt_sort( mbedtls_x509_crt *chain )
+{
+    mbedtls_x509_crt *child = chain;
+
+    while ( child != NULL)
```
Comment (gilles-peskine-arm) [inline library/x509_crt.c]: Please explain why this loop always terminates. This is not obvious since `child` can be updated in two different ways.

The loop terminates because every run through the loop always removes the current `child` from consideration.

Also, `chain` remains a pointer to the head of the list, because the head is never swapped with some other element.
